swap/src/recover.rs

```diff
@@ -0,0 +1,481 @@
+//! This module is used to attempt to recover an unfinished swap.
+//!
+//! Recovery is only supported for certain states and the strategy followed is
+//! to perform the simplest steps that require no further action from the
+//! counterparty.
+//!
+//! The quality of this module is bad because there is a lot of code
+//! duplication, both within the module and with respect to
+//! `xmr_btc/src/{alice,bob}.rs`. In my opinion, a better approach to support
+//! swap recovery would be through the `action_generator`s themselves, but this
+//! was deemed too complicated for the time being.
+
+use crate::{
+    bitcoin, monero,
+    monero::CreateWalletForOutput,
+    state::{Alice, Bob, Swap},
+};
+use anyhow::Result;
+use ecdsa_fun::{adaptor::Adaptor, nonce::Deterministic};
+use futures::{
+    future::{select, Either},
+    pin_mut,
+};
+use sha2::Sha256;
+use tracing::info;
+use xmr_btc::bitcoin::{
+    poll_until_block_height_is_gte, BroadcastSignedTransaction, TransactionBlockHeight,
+    WatchForRawTransaction,
+};
+
+pub async fn recover(
+    bitcoin_wallet: bitcoin::Wallet,
+    monero_wallet: monero::Wallet,
+    state: Swap,
+) -> Result<()> {
+    match state {
+        Swap::Alice(state) => alice_recover(bitcoin_wallet, monero_wallet, state).await,
+        Swap::Bob(state) => bob_recover(bitcoin_wallet, monero_wallet, state).await,
+    }
+}
+
+pub async fn alice_recover(
+    bitcoin_wallet: bitcoin::Wallet,
+    monero_wallet: monero::Wallet,
+    state: Alice,
+) -> Result<()> {
+    match state {
+        Alice::Handshaken(_) | Alice::BtcLocked(_) | Alice::SwapComplete => {
+            info!("Nothing to do");
+        }
+        Alice::XmrLocked(state) => {
+            info!("Monero still locked up");
+
+            let tx_cancel = bitcoin::TxCancel::new(
+                &state.tx_lock,
+                state.refund_timelock,
+                state.a.public(),
+                state.B.clone(),
+            );
+
+            info!("Checking if the Bitcoin cancel transaction has been published");
+            if bitcoin_wallet
+                .0
+                .get_raw_transaction(tx_cancel.txid())
+                .await
+                .is_err()
+            {
+                info!("Bitcoin cancel transaction not yet published");
+
+                let tx_lock_height = bitcoin_wallet
+                    .transaction_block_height(state.tx_lock.txid())
+                    .await;
+                poll_until_block_height_is_gte(
```

Comment:
```suggestion
                poll_until_block_height(
```
We can only poll until it is greater than, the name can be shorten